aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu

```diff
@@ -1223,6 +1255,39 @@ void linalg_eigh_cusolver(Tensor& eigenvalues, Tensor& eigenvectors, Tensor& inf
   }
 }
 
+void lu_solve_looped_cusolver(const Tensor& b, const Tensor& lu, const Tensor& pivots) {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(b.scalar_type(), "lu_solve_cusolver", [&] {
+    int n = cuda_int_cast(lu.size(-2), "n");
+    int nrhs = cuda_int_cast(b.size(-1), "nrhs");
+    auto batch_size = batchCount(lu);
+    auto info = at::zeros({1}, lu.options().dtype(kInt));
+    auto info_data = info.data_ptr<int>();
+    auto b_data = b.data_ptr<scalar_t>();
+    auto lu_data = lu.data_ptr<scalar_t>();
+    auto pivots_data = pivots.data_ptr<int>();
+    auto pivots_stride = pivots.size(-1);
+    auto lu_stride = matrixStride(lu);
+    auto b_stride = matrixStride(b);
+    int leading_dimension = cuda_int_cast(std::max<int>(1, n), "leading_dimension");
+
+    for (auto batch = decltype(batch_size){0}; batch < batch_size; ++batch) {
+      auto handle = at::cuda::getCurrentCUDASolverDnHandle();
```

Comment:
The handle should be inside the for loop to be able to take advantage of possible CUDA stream based parallellism.

Reply:
Currently most of our cusolver code has `handle` outside the loop. Besides, all cuda-based handles use "current stream". See https://github.com/pytorch/pytorch/blob/d75c402755af27c44c9f909dcc65e7ba35dffd91/aten/src/ATen/cuda/CusolverDnHandlePool.cpp#L43-L45

I'm not sure if multi-stream or multi-handle could increase performance. We can put that as a todo point for the next release.

Reply:
I tested with keeping the handle out and it turns that it has a minor performance improvement over keeping the handle within the loop (probably because we never really use cuSOLVER for very large batch sizes):

Reply:
@xwang233, you're right. I was thinking that `getCurrentCUDASolverDnHandle` borrows the cuda stream from the pool, but it's not. I verified with the profiler that all kernels are launched on same (default) stream.
The following code inside the for loop makes cuSOLVER use different streams:
```cpp
auto myStream = at::cuda::getStreamFromPool();
at::cuda::setCurrentCUDAStream(myStream);
auto handle = at::cuda::getCurrentCUDASolverDnHandle();
``` 
However, this specific cuSOLVER routine doesn't benefit from CUDA streams, all kernels inside the cuSOLVER are executed sequentially still. So [the note from the docs](https://docs.nvidia.com/cuda/cusolver/index.html#parallelism-with-streams) doesn't apply to all functions in cuSOLVER .
![image](https://user-images.githubusercontent.com/19621411/121055226-bf7c2100-c7c5-11eb-999d-28b68b5b76bf.png)

Reply:
@IvanYashchuk I once had a prototype of parallel cuda stream implementation during the SVD PR, see https://github.com/xwang233/pytorch/blob/445ba7783ae59fd2bd84223c82a0c9188f663ecc/aten/src/ATen/native/cuda/BatchLinearAlgebraLib.h#L25-L42

There was some discussions on the performance in that PR as well, see discussions around https://github.com/pytorch/pytorch/pull/48436#discussion_r552185100. 

This might be a large work to do, and we can put that as a todo point for the next release.

Reply:
OK so this seems to be good to go now. Let me mark this comment as resolved. I think this is ready for merge in that case.